pkg/apis/kubeone/v1beta1/types.go

```diff
@@ -0,0 +1,324 @@
+	SSHPrivateKeyFile string `json:"sshPrivateKeyFile"`
+	SSHAgentSocket    string `json:"sshAgentSocket"`
+	Bastion           string `json:"bastion"`
+	BastionPort       int    `json:"bastionPort"`
+	BastionUser       string `json:"bastionUser"`
+	Hostname          string `json:"hostname"`
+	IsLeader          bool   `json:"isLeader"`
+
+	// If not provided (i.e. nil) defaults to TaintEffectNoSchedule, with key
+	// node-role.kubernetes.io/master for control plane nodes.
+	//
+	// Explicitly empty (i.e. []corev1.Taint{}) means no taints will be applied
+	// (this is default for worker nodes).
+	Taints []corev1.Taint `json:"taints,omitempty"`
+
+	// Information populated at the runtime
+	OperatingSystem OperatingSystemName `json:"-"`
+}
+
+// OperatingSystemName defines the operating system used on instances
+type OperatingSystemName string
+
+var (
+	OperatingSystemNameUbuntu  OperatingSystemName = "ubuntu"
+	OperatingSystemNameCentOS  OperatingSystemName = "centos"
+	OperatingSystemNameCoreOS  OperatingSystemName = "coreos"
+	OperatingSystemNameFlatcar OperatingSystemName = "flatcar"
+	OperatingSystemNameDefault OperatingSystemName = ""
+)
+
+// APIEndpoint is the endpoint used to communicate with the Kubernetes API
+type APIEndpoint struct {
+	// Host is the hostname on which API is running
+	Host string `json:"host"`
+
+	// Port is the port used to reach to the API
+	Port int `json:"port"`
+}
+
+// CloudProviderSpec describes the cloud provider that is running the machines.
+// Only one cloud provider must be defined at the single time.
+type CloudProviderSpec struct {
+	External     bool              `json:"external"`
+	CloudConfig  string            `json:"cloudConfig"`
+	AWS          *AWSSpec          `json:"aws"`
+	Azure        *AzureSpec        `json:"azure"`
+	DigitalOcean *DigitalOceanSpec `json:"digitalocean"`
+	Gce          *GceSpec          `json:"gce"`
+	Hetzner      *HetznerSpec      `json:"hetzner"`
+	Openstack    *OpenstackSpec    `json:"openstack"`
+	Packet       *PacketSpec       `json:"packet"`
+	Vsphere      *VsphereSpec      `json:"vsphere"`
+	None         *NoneSpec         `json:"none"`
+}
+
+// AWSSpec defines the AWS cloud provider
+type AWSSpec struct{}
+
+// AzureSpec defines the Azure cloud provider
+type AzureSpec struct{}
+
+// DigitalOceanSpec defines the DigitalOcean cloud provider
+type DigitalOceanSpec struct{}
+
+// GceSpec defines the GCE cloud provider
+type GceSpec struct{}
+
+// HetznerSpec defines the Hetzner cloud provider
+type HetznerSpec struct{}
+
+// OpenstackSpec defines the Openstack provider
+type OpenstackSpec struct{}
+
+// PacketSepc defines the Packet cloud provider
+type PacketSpec struct{}
+
+// VsphereSpec defines the vSphere provider
+type VsphereSpec struct{}
+
+// NoneSpec defines a none provider
+type NoneSpec struct{}
+
+// VersionConfig describes the versions of components that are installed on the machines
+type VersionConfig struct {
+	Kubernetes semver.Semver `json:"kubernetes"`
```

Comment:
I'm totally no fan of including other packages as types in API. Let's handle semver at validator level.

Reply:
Sounds good. Removed the semver API and switched back to string.